src/handles/handle.rs

```diff
@@ -0,0 +1,330 @@
+        Pcr9 = ESYS_TR_PCR9,
+        Pcr10 = ESYS_TR_PCR10,
+        Pcr11 = ESYS_TR_PCR11,
+        Pcr12 = ESYS_TR_PCR12,
+        Pcr13 = ESYS_TR_PCR13,
+        Pcr14 = ESYS_TR_PCR14,
+        Pcr15 = ESYS_TR_PCR15,
+        Pcr16 = ESYS_TR_PCR16,
+        Pcr17 = ESYS_TR_PCR17,
+        Pcr18 = ESYS_TR_PCR18,
+        Pcr19 = ESYS_TR_PCR19,
+        Pcr20 = ESYS_TR_PCR20,
+        Pcr21 = ESYS_TR_PCR21,
+        Pcr22 = ESYS_TR_PCR22,
+        Pcr23 = ESYS_TR_PCR23,
+        Pcr24 = ESYS_TR_PCR24,
+        Pcr25 = ESYS_TR_PCR25,
+        Pcr26 = ESYS_TR_PCR26,
+        Pcr27 = ESYS_TR_PCR27,
+        Pcr28 = ESYS_TR_PCR28,
+        Pcr29 = ESYS_TR_PCR29,
+        Pcr30 = ESYS_TR_PCR30,
+        Pcr31 = ESYS_TR_PCR31,
+    }
+
+    impl_constant_handle_conversion_basic!(PcrHandle);
+    impl_constant_handle_conversion!(PcrHandle, ObjectHandle);
+}
+
+/// TPM Constant handle module
+pub mod tpm_constant {
+    use super::object::ObjectHandle;
+    use crate::{
+        tss2_esys::{
+            ESYS_TR, ESYS_TR_RH_ENDORSEMENT, ESYS_TR_RH_LOCKOUT, ESYS_TR_RH_NULL, ESYS_TR_RH_OWNER,
+            ESYS_TR_RH_PLATFORM, ESYS_TR_RH_PLATFORM_NV,
+        },
+        Error, Result, WrapperErrorKind,
+    };
+    use log::error;
+    use num_derive::{FromPrimitive, ToPrimitive};
+    use num_traits::{FromPrimitive, ToPrimitive};
+    use std::convert::{From, TryFrom};
+    /// TpmConstant Handle
+    ///
+    /// Represents an esys handle for
+    /// tpm constant resource handles.
+    #[derive(FromPrimitive, ToPrimitive, Debug, Copy, Clone, PartialEq, Eq)]
+    #[repr(u32)]
+    pub enum TpmConstantHandle {
+        Owner = ESYS_TR_RH_OWNER,
+        Null = ESYS_TR_RH_NULL,
+        Lockout = ESYS_TR_RH_LOCKOUT,
+        Endorsement = ESYS_TR_RH_ENDORSEMENT,
+        Platform = ESYS_TR_RH_PLATFORM,
+        PlatformNv = ESYS_TR_RH_PLATFORM_NV,
+    }
+
+    impl_constant_handle_conversion_basic!(TpmConstantHandle);
+    impl_constant_handle_conversion!(TpmConstantHandle, ObjectHandle);
+}
+
+///
+/// Macro for implmeneting conversion between handles.
+///
+macro_rules! impl_handle_conversion {
+    ($handle_type_self:ident, $handle_type_other:ident) => {
+        impl From<$handle_type_self> for $handle_type_other {
+            fn from(handle_self: $handle_type_self) -> $handle_type_other {
+                $handle_type_other::from(ESYS_TR::from(handle_self.value))
+            }
+        }
+
+        impl From<$handle_type_other> for $handle_type_self {
+            fn from(handle_other: $handle_type_other) -> $handle_type_self {
+                $handle_type_self {
+                    value: handle_other.value(),
+                }
+            }
+        }
+    };
+}
+
+/// Auth handle module
+pub mod auth {
+    use super::object::ObjectHandle;
+    impl_basic_handle!(
+        /// Auth handle
+        ///
+        /// Represents an esys handle to resources
+        /// that can be used for authentication.
+        AuthHandle
```

Comment:
I'm very confused as to what this type is supposed to represent. I'm not sure you've intended it the same way `ESYS_TR_RH_AUTH` is defined in the ESAPI spec

Reply:
Nope I have not. An AuthHandle is a handle to anything that have metadata in it that can be used for authorization.
So for example when you do
```rust
context.set_sessions((ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE));
let nv_index_handle = context.nv_define_space(NvAuthorization::Owner, Some(&Auth::try_from(vec![0,1,2,3])?), &nv_public).unwrap();
```
Then the auth is stored in esys and is associated with the handle returned from esapi. And the nv index handle can then be used as an AuthHandle.


And I was a little bit unsure how to interpret that range of values in the specification:
> #define ESYS_TR_RH_AUTH_FIRST 0x110U
> #define ESYS_TR_RH_AUTH(x) (ESYS_TR_RH_AUTH_FIRST + (ESYS_TR)(x))

Didn't understand if that was something I had to bother with or not.

Reply:
So those would have been added to the TpmConstantsHandle(terrible name I know but it is what esapi spec calls them).